File: 1123.Lowest Common Ancestor of Deepest Leaves.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * Finds LCA of deepest leaves using DFS with global maxDepth tracking
 * 
 * #intuition
 * - Track global maxDepth to identify deepest level
 * - Node is LCA when both subtrees reach maxDepth
 * - Update LCA when finding equal depths at maxDepth
 * 
 * #approach
 * 1. Use DFS with depth parameter to track current depth
 * 2. Update maxDepth when reaching deeper nodes
 * 3. Compare left and right subtree depths
 * 4. Set node as LCA when both subtrees reach maxDepth
 * 5. Return maximum depth of current subtree
 * 
 * #complexity
 * Time: O(n) where n is number of nodes, single DFS pass
 * Space: O(h) where h is tree height, recursion stack only
 * 
 * @param {TreeNode} root - Root node of the binary tree
 * @return {TreeNode} - Lowest common ancestor of deepest leaves
 */
var lcaDeepestLeaves = function(root) {
    let maxDepth = -1;
    let lca = null;
    
    function dfs(node, depth) {
        if (!node) {
            return depth - 1;
        }
        
        // Update maxDepth when reaching deeper level
        if (depth > maxDepth) {
            maxDepth = depth;
        }
        
        const leftDepth = dfs(node.left, depth + 1);
        const rightDepth = dfs(node.right, depth + 1);
        
        // Update LCA when both subtrees reach maxDepth
        if (leftDepth === rightDepth && leftDepth === maxDepth) {
            lca = node;
        }
        
        return Math.max(leftDepth, rightDepth);
    }
    
    dfs(root, 0);
    return lca;
};

File: 1123.Lowest Common Ancestor of Deepest Leaves.rs
use std::cell::RefCell;
use std::cmp::Ordering;
use std::rc::Rc;

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

impl Solution {
    /// Finds LCA of deepest leaves using DFS with depth tracking.
    ///
    /// # Intuition
    /// The LCA of deepest leaves is the node where both subtrees reach the
    /// maximum depth. If one subtree is deeper, propagate that result upward.
    ///
    /// # Approach
    /// DFS returning `(depth, lca_candidate)`. If left and right depths are
    /// equal, the current node is the LCA. Otherwise return the deeper side.
    ///
    /// # Complexity
    /// - Time: O(n)
    /// - Space: O(h) recursion stack
    pub fn lca_deepest_leaves(
        root: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        Self::dfs(&root).1
    }

    fn dfs(node: &Option<Rc<RefCell<TreeNode>>>) -> (u32, Option<Rc<RefCell<TreeNode>>>) {
        node.as_ref().map_or((0, None), |n| {
            let borrowed = n.borrow();
            let (left_depth, left_lca) = Self::dfs(&borrowed.left);
            let (right_depth, right_lca) = Self::dfs(&borrowed.right);

            match left_depth.cmp(&right_depth) {
                Ordering::Equal => (left_depth + 1, Some(n.clone())),
                Ordering::Greater => (left_depth + 1, left_lca),
                Ordering::Less => (right_depth + 1, right_lca),
            }
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn build_tree(values: Vec<Option<i32>>) -> Option<Rc<RefCell<TreeNode>>> {
        if values.is_empty() || values[0].is_none() {
            return None;
        }

        let root = Rc::new(RefCell::new(TreeNode::new(values[0].unwrap())));
        let mut queue = std::collections::VecDeque::with_capacity(values.len());
        queue.push_back(root.clone());

        let mut i = 1;
        while i < values.len() && !queue.is_empty() {
            let node = queue.pop_front().unwrap();

            if i < values.len() && values[i].is_some() {
                let left = Rc::new(RefCell::new(TreeNode::new(values[i].unwrap())));
                node.borrow_mut().left = Some(left.clone());
                queue.push_back(left);
            }
            i += 1;

            if i < values.len() && values[i].is_some() {
                let right = Rc::new(RefCell::new(TreeNode::new(values[i].unwrap())));
                node.borrow_mut().right = Some(right.clone());
                queue.push_back(right);
            }
            i += 1;
        }

        Some(root)
    }

    #[test]
    fn test_basic_tree() {
        // Tree: [3,5,1,6,2,0,8,null,null,7,4]
        // Deepest leaves: 7, 4
        // LCA: 2
        let root = build_tree(vec![
            Some(3),
            Some(5),
            Some(1),
            Some(6),
            Some(2),
            Some(0),
            Some(8),
            None,
            None,
            Some(7),
            Some(4),
        ]);
        let result = Solution::lca_deepest_leaves(root);
        assert!(result.is_some());
        assert_eq!(result.unwrap().borrow().val, 2);
    }

    #[test]
    fn test_single_node() {
        // Tree: [1]
        // Deepest leaves: 1
        // LCA: 1
        let root = build_tree(vec![Some(1)]);
        let result = Solution::lca_deepest_leaves(root);
        assert!(result.is_some());
        assert_eq!(result.unwrap().borrow().val, 1);
    }

    #[test]
    fn test_left_skewed() {
        // Tree: [1,2,null,3]
        // Deepest leaf: 3
        // LCA: 3
        let root = build_tree(vec![Some(1), Some(2), None, Some(3)]);
        let result = Solution::lca_deepest_leaves(root);
        assert!(result.is_some());
        assert_eq!(result.unwrap().borrow().val, 3);
    }

    #[test]
    fn test_two_deepest_leaves() {
        // Tree: [1,2,3,4,5]
        // Deepest leaves: 4, 5
        // LCA: 2
        let root = build_tree(vec![Some(1), Some(2), Some(3), Some(4), Some(5)]);
        let result = Solution::lca_deepest_leaves(root);
        assert!(result.is_some());
        assert_eq!(result.unwrap().borrow().val, 2);
    }

    #[test]
    fn test_empty_tree() {
        let result = Solution::lca_deepest_leaves(None);
        assert!(result.is_none());
    }

    #[test]
    fn test_all_leaves_same_depth() {
        // Tree: [1,2,3]
        // Deepest leaves: 2, 3
        // LCA: 1
        let root = build_tree(vec![Some(1), Some(2), Some(3)]);
        let result = Solution::lca_deepest_leaves(root);
        assert!(result.is_some());
        assert_eq!(result.unwrap().borrow().val, 1);
    }
}
